From 16029c8d0582452adf0e1a76933b98b44d5ca885 Mon Sep 17 00:00:00 2001
From: Junping Du <junping_du@apache.org>
Date: Sun, 15 Feb 2015 06:46:32 -0800
Subject: [PATCH 1235/1280] YARN-2749. Fix some testcases from
 TestLogAggregationService fails in trunk.
 (Contributed by Xuan Gong) (cherry picked from
 commit ab0b958a522d502426b91b6e4ab6dd29caccc372)

(cherry picked from commit eee89bd18c0c82f14f76058b0000093a1ecce205)

Change-Id: I78ec2faf5698a634e7b42003453eea38ca4ab974
(cherry picked from commit 8eb01bfb8236be5673c02d1e4aecd38fc190cc8b)
---
 .../logaggregation/AppLogAggregatorImpl.java       |   15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/logaggregation/AppLogAggregatorImpl.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/logaggregation/AppLogAggregatorImpl.java
index d8e4899..ce2ec7b 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/logaggregation/AppLogAggregatorImpl.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/logaggregation/AppLogAggregatorImpl.java
@@ -121,6 +121,8 @@
   private final int retentionSize;
   private final long rollingMonitorInterval;
   private final NodeId nodeId;
+  // This variable is only for testing
+  private final AtomicBoolean waiting = new AtomicBoolean(false);
 
   private boolean renameTemporaryLogFileFailed = false;
 
@@ -432,6 +434,7 @@ private void doAppLogAggregation() {
     while (!this.appFinishing.get() && !this.aborted.get()) {
       synchronized(this) {
         try {
+          waiting.set(true);
           if (this.rollingMonitorInterval > 0) {
             wait(this.rollingMonitorInterval * 1000);
             if (this.appFinishing.get() || this.aborted.get()) {
@@ -548,7 +551,19 @@ public synchronized void abortLogAggregation() {
 
   @Private
   @VisibleForTesting
+  // This is only used for testing.
+  // This will wake the log aggregation thread that is waiting for
+  // rollingMonitorInterval.
+  // To use this method, make sure the log aggregation thread is running
+  // and waiting for rollingMonitorInterval.
   public synchronized void doLogAggregationOutOfBand() {
+    while(!waiting.get()) {
+      try {
+        wait(200);
+      } catch (InterruptedException e) {
+        // Do Nothing
+      }
+    }
     LOG.info("Do OutOfBand log aggregation");
     this.notifyAll();
   }
-- 
1.7.9.5

